EIP712/Implementation/EIP712.cs

```diff
@@ -0,0 +1,203 @@
+        /// <summary>
+        /// Implements hashStruct from EIP712
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="structure"></param>
+        /// <returns></returns>
+        private static byte[] HashStruct<T>(T structure) where T : class
+        {
+            Type structType = structure.GetType();
+
+            // Get all properties on which StructTypeAttribute is applied and order them by "Order" property
+            Tuple<PropertyInfo, MemberAttribute>[] props = structType.GetTypeInfo().DeclaredProperties.
+                Where(prop => prop.CustomAttributes.Any(attr => attr.AttributeType == typeof(MemberAttribute))).
+                Select(prop => Tuple.Create(prop, prop.GetCustomAttribute<MemberAttribute>())).
+                OrderBy(propAttrPair => propAttrPair.Item2.Order).ToArray();
+
+            byte[] typeHash = CalculateTypeHash(structure, props);
+            byte[] encodedData = EncodeData(structure, props);
+            byte[] merged = ByteUtil.Merge(typeHash, encodedData);
+            return _keccak.CalculateHash(merged);
+        }
+
+        private static byte[] CalculateTypeHash<T>(T structure,
+            Tuple<PropertyInfo, MemberAttribute>[] props)
+        {
+
+            Type structType = structure.GetType();
+            StructNameAttribute nameAttr = structType.GetTypeInfo().
+                GetCustomAttribute<StructNameAttribute>();
+            string encodedType = $"{(nameAttr == null ? structType.Name : nameAttr.Name)}(";
+
+            string nameTypeConcatenated = props.Aggregate(string.Empty, (accumulated, prop) =>
+            {
+
+                // If values is null do not encode type
+                if (prop.Item1.GetValue(structure) == null)
+                    return accumulated;
+
+                string prefix = accumulated == string.Empty ? string.Empty : ",";
+                // TODO: Make property name encoding configurable
+                string nameType = $"{prefix}{prop.Item2.AbiType} {prop.Item1.Name.ToCamelCase()}";
+
+                return accumulated + nameType;
+            });
+
+            encodedType += $"{nameTypeConcatenated})";
+
+            return _keccak.CalculateHash(Encoding.UTF8.GetBytes(encodedType));
+        }
+
+        private static byte[] EncodeData<T>(T structure, Tuple<PropertyInfo, MemberAttribute>[] props)
+        {
+            byte[] result = new byte[0];
+            byte[] part = null;
+            object val = null;
+
+            foreach (var prop in props)
+            {
+
+                // TODO: Support more types
+                // TODO: Implement proper error handling
+
+                val = prop.Item1.GetValue(structure);
+                string abiType = prop.Item2.AbiType;
+                PropertyInfo propInfo = prop.Item1;
+
+                // Do not encode properties with null values
+                if (val == null)
+                    continue;
+
+                switch (abiType)
```

Comment:
Most of these types are encoded the same (`uint*`, `bytes*`, most static sized types), so maybe instead of a exact `switch` statement try the new pattern matching from C# 7 or a `string.Contains` logic (to stay DRY)